document-portal/file-transfer.c

```diff
@@ -0,0 +1,544 @@
+  g_hash_table_steal (transfers, transfer->key);
+  G_UNLOCK (transfers);
+
+  g_idle_add (stop, transfer);
+}
+
+static void
+file_transfer_add_file (FileTransfer *transfer,
+                        const char *path,
+                        struct stat *parent_st_buf)
+{
+  ExportedFile *file;
+
+  file = g_new (ExportedFile, 1);
+  file->path = g_strdup (path);
+  file->parent_dev = parent_st_buf->st_dev;
+  file->parent_ino = parent_st_buf->st_ino;
+
+  g_ptr_array_add (transfer->files, file);
+}
+
+static char **
+file_transfer_execute (FileTransfer *transfer,
+                       XdpAppInfo *target_app_info,
+                       GError **error)
+{
+  guint32 flags;
+  DocumentPermissionFlags perms;
+  const char *target_app_id;
+  int n_fds;
+  g_autofree int *fds = NULL;
+  int i;
+  g_auto(GStrv) ids = NULL;
+  char **files = NULL;
+
+  g_debug ("retrieve %d files for %s from file transfer owned by '%s' (%s)",
+           transfer->files->len,
+           xdp_app_info_get_id (target_app_info),
+           xdp_app_info_get_id (transfer->app_info),
+           transfer->sender);
+
+  /* if the target is unsandboxed, just return the files as-is */
+  if (xdp_app_info_is_host (target_app_info))
+    {
+      files = g_new (char *, transfer->files->len + 1);
+      for (i = 0; i < transfer->files->len; i++)
+        {
+          ExportedFile *file = (ExportedFile*)g_ptr_array_index (transfer->files, i);
+          files[i] = g_strdup (file->path);
+        }
+      files[i] = NULL;
+      return files;
+    }
+
+  flags = DOCUMENT_ADD_FLAGS_REUSE_EXISTING | DOCUMENT_ADD_FLAGS_AS_NEEDED_BY_APP;
+
+  perms = DOCUMENT_PERMISSION_FLAGS_READ;
+  if (transfer->writable)
+    perms |= DOCUMENT_PERMISSION_FLAGS_WRITE;
+
+  target_app_id = xdp_app_info_get_id (target_app_info);
+
+  n_fds = transfer->files->len;
+  fds = g_new (int, n_fds);
+  for (i = 0; i < n_fds; i++)
+    {
+      g_autofree char *dirname = NULL;
+      xdp_autofd int dir_fd = -1;
+      struct stat parent_st_buf;
+      ExportedFile *file;
+
+      file = (ExportedFile*)g_ptr_array_index (transfer->files, i);
+      dirname = g_path_get_dirname (file->path);
+      dir_fd = open (dirname, O_CLOEXEC | O_PATH);
+
+      /* Check that the parent dir hasn't changed since between the
+       * time that the file was added to the transfer and the time
+       * that the transfer is executed, to prevent a symlink attack.
+       */
+      if (dir_fd < 0 || fstat (dir_fd, &parent_st_buf) < 0 ||
+          parent_st_buf.st_dev != file->parent_dev ||
+          parent_st_buf.st_ino != file->parent_ino)
+        {
+          int j;
+          for (j = 0; j < i; j++)
+            close (fds[j]);
+          g_set_error (error,
+                       XDG_DESKTOP_PORTAL_ERROR,
+                       XDG_DESKTOP_PORTAL_ERROR_NOT_ALLOWED,
+                       "Can't export file");
+          return NULL;
+        }
+
+      fds[i] = open (file->path, O_PATH | O_CLOEXEC);
```

Comment:
This is still racy. The same kind of attack can happen between the fstat() and the time when document_add_full does a stat on the parent dir, because *those* are the parent ino+dev that will be stored in the document portal. To fix the race i think you need to pass  file->parent_dev/ino into document_add_full and ensure that those match exactly what is being stored in the document db.